pkg/plugins/pluginrequestmeta/plugin_request_meta.go

```diff
@@ -0,0 +1,45 @@
+package pluginrequestmeta
+
+import (
+	"context"
+	"errors"
+)
+
+// StatusSource is an enum-like string value representing the source of a
+// plugin query data response status code
+type StatusSource string
+
+const (
+	StatusSourceNone       StatusSource = ""
+	StatusSourcePlugin     StatusSource = "plugin"
+	StatusSourceDownstream StatusSource = "downstream"
+)
+
+type statusSourceCtxKey struct{}
+
+// StatusSourceFromContext returns the plugin request status source stored in the context.
+// If no plugin request status source is stored in the context, [StatusSourceNone] is returned.
+func StatusSourceFromContext(ctx context.Context) StatusSource {
+	value, ok := ctx.Value(statusSourceCtxKey{}).(*StatusSource)
+	if ok {
+		return *value
+	}
+	return StatusSourceNone
```

Comment:
Need to decide if we want to return empty source or default to plugin or downstream. I think an empty string might be confusing to add to a prometheus metric label so I would suggest defaulting to plugin source

Reply:
Yeah this is technically different than the prometheus label. This is the value stored in the context. In our case, the error_source is set to "plugin" by default at the beginning of the `MetricsMiddleware`:

https://github.com/grafana/grafana/blob/ee7b4985ef18126d299292286dbe9d3eb1fb8caa/pkg/services/pluginsintegration/clientmiddleware/metrics_middleware.go#L164

so `StatusSourceNone` can only be returned in a `context.Context` that's not plugin-related (or before the metrics middleware).

I think it would also makes sense if we return `StatusSourcePlugin` by default, and can be less confusing, so either work for me :smile: